esb: Expand ESB_NEVER_DISABLE_TX mode

maje-emb opened the pull request:
This change allows the ESB_NEVER_DISABLE_TX mode to be used for
all packets, regardless of whether they require an ACK or not

For non-ACK packets, the radio transitions to the TXIDLE state
after transmission instead of DISABLED.

For ACK packets, the behavior remains unchanged
the radio transitions to the DISABLED state after transmission.

This modification ensures correct protocol behavior when
the packet type (ACK or non-ACK) is changed at runtime
with the ESB_NEVER_DISABLE_TX option enabled.

Renamed the Kconfig option CONFIG_ESB_NEVER_DISABLE_TX to
CONFIG_ESB_KEEP_TX_IDLE to better reflect its behavior.

test_proprietary: PR-99

NordicBuilder commented:
## CI Information

<sup>_To view the history of this post, clich the 'edited' button above_</sup>
Build number: 2
### Inputs:
#### Sources:
<details><summary> more details</summary>
</details>

#### Github labels
| Enabled | Name | Description |
| -- | -- | -- |
|| ci-disabled | _Disable the ci execution_ | 
|| ci-all-test | _Run all of ci, no test spec filtering will be done_ | 
|| ci-force-downstream | _Force execution of downstream even if twister fails_ | 
|| ci-run-twister | _Force run twister_ | 
|| ci-run-zephyr-twister | _Force run zephyr twister_ | 

<details><summary> List of changed files detected by CI (0)</summary>

```
```
</details>

### Outputs:

#### Toolchain
Version: 
Build docker image: 

#### Test Spec & Results: :white_check_mark: Success; :x: Failure; :orange_circle: Queued; :yellow_circle: Progress; :white_medium_square: Skipped; :warning: Quarantine
- :x: Toolchain
- :x: Build twister
- :x: Integration tests



*Note: This message is automatically posted and updated by the CI*

peknis commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -143,7 +143,13 @@ DECT NR+
 -------------------------
 
-|no_changes_yet_note|
+* Added:
+
+  * :kconfig:option:`CONFIG_ESB_NEVER_DISABLE_TX` Kconfig option support for packets with ACK.
+
+* Updated:
+
+  * Renamed :kconfig:option:`CONFIG_ESB_NEVER_DISABLE_TX` to :kconfig:option:`CONFIG_ESB_KEEP_TX_IDLE`.
```
This does not make much sense. You first add CONFIG_ESB_NEVER_DISABLE_TX and then rename it.

maje-emb replied:
I updated the changelog.

peknis commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -143,7 +143,10 @@ DECT NR+
-|no_changes_yet_note|
+* Updated:
+
+  * Renamed :kconfig:option:`CONFIG_ESB_NEVER_DISABLE_TX` Kconfig option to :kconfig:option:`CONFIG_ESB_KEEP_TX_IDLE` Kconfig option.
+  * Extended :kconfig:option:`CONFIG_ESB_KEEP_TX_IDLE` Kconfig option to apply also to packets with acknowledgment (ACK).
```
```suggestion
  * By renaming the ``CONFIG_ESB_NEVER_DISABLE_TX`` Kconfig option to :kconfig:option:`CONFIG_ESB_KEEP_TX_IDLE`.
  * By extending the :kconfig:option:`CONFIG_ESB_KEEP_TX_IDLE` Kconfig option to apply also to packets with acknowledgment (ACK).
```

grochu commented on subsys/esb/esb.c:
```diff
@@ -1366,12 +1336,16 @@ static void on_radio_disabled_tx(void)
 	nrf_radio_packetptr_set(NRF_RADIO, rx_payload_buffer);
 	if (fast_switching) {
 		nrf_radio_task_trigger(NRF_RADIO, NRF_RADIO_TASK_RXEN);
+		while (!nrf_radio_event_check(NRF_RADIO, NRF_RADIO_EVENT_READY)) {
```
Is this wait required here? The next possible event is reception of ACK or timeout, so no need to block here.

maje-emb replied:
Right, it is not needed.

grochu commented on subsys/esb/esb.c:
```diff
@@ -1402,9 +1376,33 @@ static void on_radio_disabled_tx_wait_for_ack(void)
 		} else {
+			if (IS_ENABLED(CONFIG_ESB_KEEP_TX_IDLE)) {
+				nrf_radio_task_trigger(NRF_RADIO, NRF_RADIO_TASK_TXEN);
+				while (!nrf_radio_event_check(NRF_RADIO, NRF_RADIO_EVENT_READY)) {
```
Please add a comment for clarity why there is a need to block here waiting for the Ready event.

maje-emb replied:
This block of code was removed. For the ACK package, the radio goes to the DISABLED state.

grochu reviewed (changes requested):
Thank you for this update. Apart of adding a useful protocol mode it cleans up the radio shorts setting and some other sections of code.
I have a couple of minor comments though.

maje-emb commented:
> Hi @maje-emb , I have tested your changes to check if they were working fine on my end. Unfortunately, I still see some issues. This is how I tested it:
> 
> 1. I updated all esb files from the PR and I am using the esb_ptx example
> 2. I have added the flag CONFIG_ESB_KEEP_TX_IDLE=y to prj.conf of this example
> 3. In main.c from this example, I have added some simple code to toggle ack and nack packets with a counter, to check that both can be sent. The counter toggles between ack and nack packets, so sometimes ack is sent and sometimes nack is sent.
> 4. I can see that the code hangs after moving from noack=true to noack=false. I think this is the same that happened before.
> 
> Can you try to reproduce this? Maybe I am missing something?

Hi @dblascoserrano, thanks for your feedback.
For the PTX sample, modified in this way:
```
	while (1) {
		if (ready) {
			cnt++;
			tx_payload.noack = false;
			if (cnt == 4) {
				tx_payload.noack = true;
				cnt = 0;
			}
			ready = false;
			esb_flush_tx();
			leds_update(tx_payload.data[1]);

			err = esb_write_payload(&tx_payload);
			if (err) {
				LOG_ERR("Payload write failed, err %d", err);
			}
			tx_payload.data[1]++;
		}
		k_sleep(K_MSEC(100));
	}
```
I do not see the code hanging.
Could you please post your modification?

maje-emb commented:
> Hi @maje-emb , I have tested your changes to check if they were working fine on my end. Unfortunately, I still see some issues. This is how I tested it:
> 
> 1. I updated all esb files from the PR and I am using the esb_ptx example
> 2. I have added the flag CONFIG_ESB_KEEP_TX_IDLE=y to prj.conf of this example
> 3. In main.c from this example, I have added some simple code to toggle ack and nack packets with a counter, to check that both can be sent. The counter toggles between ack and nack packets, so sometimes ack is sent and sometimes nack is sent.
> 4. I can see that the code hangs after moving from noack=true to noack=false. I think this is the same that happened before.
> 
> Can you try to reproduce this? Maybe I am missing something?

Hi @dblascoserrano Thank you for your feedback.
It seems that the radio interrupts were misaligned.
Could you please check your sample again?

dblascoserrano commented:
> > Hi @maje-emb , I have tested your changes to check if they were working fine on my end. Unfortunately, I still see some issues. This is how I tested it:
> > 
> > 1. I updated all esb files from the PR and I am using the esb_ptx example
> > 2. I have added the flag CONFIG_ESB_KEEP_TX_IDLE=y to prj.conf of this example
> > 3. In main.c from this example, I have added some simple code to toggle ack and nack packets with a counter, to check that both can be sent. The counter toggles between ack and nack packets, so sometimes ack is sent and sometimes nack is sent.
> > 4. I can see that the code hangs after moving from noack=true to noack=false. I think this is the same that happened before.
> > 
> > Can you try to reproduce this? Maybe I am missing something?
> 
> Hi @dblascoserrano Thank you for your feedback. It seems that the radio interrupts were misaligned. Could you please check your sample again?

Hi @maje-emb ,

After your last change, it seems the code does not get stuck anymore and the transmission is working well! However, I see something I was not expecting:
Now, if the packet is with ACK, the radio will turn on as soon as the ACK is received (even if there is no packet to be sent afterwards). You can see that simply by using your esb_ptx example with the config flag enabled. I suppose this is not desired, the behaviour should only change in nACK packets, right?

maje-emb commented:
> Now, if the packet is with ACK, the radio will turn on as soon as the ACK is received (even if there is no packet to be sent afterwards). You can see that simply by using your esb_ptx example with the config flag enabled. I suppose this is not desired, the behaviour should only change in nACK packets, right?

Now when CONFIG_ESB_KEEP_TX_IDLE is enabled then after sending a packet with ACK the radio goes to TXIDLE state instead of DISABLE. (even if there is no another packet to send).
This keeps the radio in a ready to transmit state.

Do I understand correctly, the expectation is that after sending:
- ACK packet, the radio goes to DISABLED state
- nACK packet, the radio goes to TXIDLE state

dblascoserrano commented:
> > Now, if the packet is with ACK, the radio will turn on as soon as the ACK is received (even if there is no packet to be sent afterwards). You can see that simply by using your esb_ptx example with the config flag enabled. I suppose this is not desired, the behaviour should only change in nACK packets, right?
> 
> Now when CONFIG_ESB_KEEP_TX_IDLE is enabled then after sending a packet with ACK the radio goes to TXIDLE state instead of DISABLE. (even if there is no another packet to send). This keeps the radio in a ready to transmit state.
> 
> Do I understand correctly, the expectation is that after sending:
> 
> * ACK packet, the radio goes to DISABLED state
> * nACK packet, the radio goes to TXIDLE state

Yes, exactly. After ACK, the behaviour is the same as before. After nACK, the radio stays ON.

maje-emb commented:
> > > Now, if the packet is with ACK, the radio will turn on as soon as the ACK is received (even if there is no packet to be sent afterwards). You can see that simply by using your esb_ptx example with the config flag enabled. I suppose this is not desired, the behaviour should only change in nACK packets, right?
> > 
> > 
> > Now when CONFIG_ESB_KEEP_TX_IDLE is enabled then after sending a packet with ACK the radio goes to TXIDLE state instead of DISABLE. (even if there is no another packet to send). This keeps the radio in a ready to transmit state.
> > Do I understand correctly, the expectation is that after sending:
> > 
> > * ACK packet, the radio goes to DISABLED state
> > * nACK packet, the radio goes to TXIDLE state
> 
> Yes, exactly. After ACK, the behaviour is the same as before. After nACK, the radio stays ON.

Thanks for the clarification.
I have updated the PR.
I'll update the documentation.

maje-emb commented on doc/nrf/protocols/esb/index.rst:
```diff
@@ -333,10 +333,9 @@ Experimental feature: Never disable transmission stage
+Once the packet has been transmitted, the radio peripheral remains in TXIDLE state instead of TXDISABLE when transmission is pending.
 However, this process consumes more energy, because the radio transmitter stage remains enabled when transmission is taking place.
-In this mode, the :c:member:`esb_config.retransmit_delay` field specifies the delay between consecutive packet transmissions from the TX FIFO.
+For packages without acknowledgment, the :c:member:`esb_config.packet_delay` field specifies the delay between consecutive packet transmissions from the TX FIFO.
```
Yes. Good point.

peknis commented on subsys/esb/esb_dppi.c:
```diff
@@ -40,8 +42,17 @@ void esb_ppi_for_txrx_set(bool rx, bool timer_start)
-	nrf_radio_subscribe_set(NRF_RADIO, rx ? NRF_RADIO_TASK_RXEN : NRF_RADIO_TASK_TXEN,
-				egu_ramp_up);
+	if (state == NRF_RADIO_STATE_TXIDLE || state == NRF_RADIO_STATE_TX) {
+		/* When CONFIG_ESB_KEEP_TX_IDLE is enabled the radio is running,
+		 * so we need to trigger NRF_RADIO_TASK_START.
```
```suggestion
		/* When CONFIG_ESB_KEEP_TX_IDLE is enabled, the radio is running.
		 * NRF_RADIO_TASK_START needs to be triggered.
```
Not really our cup of tea to fix the comments in the code. However decided to leave a minor suggestion.
It is not a good idea to talk about "we".

maje-emb replied:
Thanks for your comment.

peknis reviewed (approved):
Left a minor nit.

carlescufi reviewed (changes requested):
Blocked at the request of Svein.

peknis commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -143,7 +143,10 @@ DECT NR+
+* Updated:
+
+  * By renaming the ``CONFIG_ESB_NEVER_DISABLE_TX`` Kconfig option to :kconfig:option:`CONFIG_ESB_KEEP_TX_IDLE`.
+  * By enabling runtime switching between ACK and non-ACK packets with :kconfig:option:`CONFIG_ESB_KEEP_TX_IDLE` Kconfig option enabled.
```
```suggestion
  * By enabling runtime switching between ACK and non-ACK packets using the :kconfig:option:`CONFIG_ESB_KEEP_TX_IDLE` Kconfig option.
```

NordicBuilder commented:
This pull request has been marked as stale because it has been open (more than) 30 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this pull request will automatically be closed in 7 days. Note, that you can always re-open a closed pull request at any time.
